# Non-canonical DNA in human and other ape telomere-to-telomere genomes
<https://www.biorxiv.org/content/10.1101/2024.09.02.610891v2>  
Linnéa Smeds, Kaivan Kamali, Iva Kejnovská, Eduard Kejnovský, Francesca Chiaromonte, Kateryna D. Makova  
Department of Biology, Penn State University, University Park, PA 16802  
Correspondence to Kateryna Makova (kdm16@psu.edu)  

***

## *THIS PAGE IS STILL UNDER CONSTRUCTION*

### This repository includes the following files and directories:

#### **general_commands.sh**
Includes a list of software, commands for creating the bed files, generate statistics and find overlaps between the motif types. 

#### **density_commands.sh**
Commands for calculating non-B DNA motif density and GC content in 1-00kb windows.

#### **new_sequence_commands.sh**
Analysis of new sequence in the T2T assemblies vs. old sequence (non T2T genomes), including scripts to align new vs old. 

#### **repeat_commands.sh**
Commands for enrichment analysis in repeats (including paths to repeat annotations). Also contains the detailed analysis of the satellites Walusat, LSAU and SST1. 

#### **functional_commands.sh**
Commands for analysis of functional regions.

#### **methylation_commands.sh**
Commands for methylation analysis of certain repeats and satellites.

#### **centromere_commands.sh**
Analysis of centromeres, including centromeric satellites, CENP-B and SF analysis.

#### **circos_commands.sh**
Scripts for generating input files for circos plots.

#### **circos/**
Examples of circos scripts for human (including all configuration files and nonB data files).

#### **python/**
All in-house python scripts used in the study. 

#### **R/**
All R scripts to generate figures and statistics.

