#################################################################################
### R code for plotting enrichments in repeats (detail for human, summary for 
### all species) and methylation scores in satellites/G4s in human.
### written by Linn√©a Smeds Dec-2024

# Note: This plot is combined manually with Part D (CD analysis) generated by 
# I. Kejnovska and E. Kejnovsky
################################################################################
# Setting up, loading R libraries
rm(list=ls())
require(tidyverse)
require(patchwork)  

# Parameters 
sp="human"
threshold <- 50000    # Example: 50000
plotdir="plots/"


# Input files
repeat_file <- paste("repeats/",sp,"_genome_repeat_enrichment_renamed.tsv", sep="")
lengths_file <- paste("repeats/",sp,"/genome_repeat_lengths_renamed.txt", sep="")
HG002_rep<-paste("methylation/",sp,"/repeats_with_scores_HG002_shortnames.bed", sep="")
HG002_g4<-paste("methylation/",sp,"/overlap/G4s_in_roi_with_scores_HG002_shortnames.bed", sep="")
chm13_rep<-paste("methylation/",sp,"/repeats_with_scores_chm13_shortnames.bed", sep="")
chm13_g4<-paste("methylation/",sp,"/overlap/G4s_in_roi_with_scores_chm13_shortnames.bed", sep="")
groupRep_file <- "repeats/7sp_group_enrichment.tsv"



################################################################################
# Processing files, convert to tibbles 

### REPEAT ENRICHMENT DATA
# Read files and pivot table to put motifs in single column 
lentib <- lengths_file %>% read.table(header=FALSE) %>% rename(Repeat=V1, bp=V2)
tib <- repeat_file %>% read.table(header=TRUE) %>% as_tibble() %>% 
      left_join(lentib) %>% pivot_longer(-c(Repeat,bp), names_to="Motif") %>%
      mutate(PrintLen=paste("(",sprintf("%.2f",bp/1000000)," Mb)", sep="")) %>% 
      mutate(PrintName=paste(Repeat," (",sprintf("%.2f",bp/1000000),"Mb)", sep="")) %>%
      mutate(Motif = str_replace_all(Motif, c('GQ' = 'G4'))) %>%
      distinct(bp, Motif, .keep_all = TRUE) %>%
      mutate(Mark=if_else((Repeat=="Retroposon/SVA" | Repeat=="rRNA" |Repeat=="Satellite/acro_ACRO1" |
                          Repeat=="Satellite/centr_GSAT" | Repeat=="Satellite/centr_GSATII" |
                          Repeat=="Satellite/centr_GSATX" | Repeat=="Satellite/centr_SST1" | 
                          Repeat=="Satellite/subtelo_TAR1" | Repeat=="Satellite_HSAT5" |  
                            Repeat=="Satellite_LSAU"  | Repeat=="Satellite_MSR1" | 
                            Repeat=="Satellite_SAT*6*3554" | Repeat=="Satellite_TAR1"), "Yes", NA)) %>%  
      mutate(Class=case_when(str_detect(Repeat, "DNA") ~ "TEs",
                             str_detect(Repeat, "LINE") ~ "TEs",
                             str_detect(Repeat, "SINE") ~ "TEs",
                             str_detect(Repeat, "LTR") ~ "TEs",
                             str_detect(Repeat, "Helitron") ~ "TEs",
                             str_detect(Repeat, "Retroposon") ~ "TEs",
                             str_detect(Repeat, "Satellite") ~ "Satellites",
                             str_detect(Repeat, "RNA") ~ "RNA",
                             TRUE ~ "Other")) %>%
  mutate(PrintName = sub("SAT", "Sat", PrintName))
                                                                   
REPDATA <- tib %>% filter(bp>threshold) %>% mutate(logval=log2(value))
REPDATA$Motif <- factor(REPDATA$Motif, levels=c("APR", "DR", "STR", "IR", "MR", "TRI", "G4", "Z"))
REPDATA$Class <- factor(REPDATA$Class, levels=c("TEs", "Satellites", "RNA", "Other"))

### METHYLATION DATA IN SATELLITES & G4s
# Tibble with scores - full repeat
rep_tib_HG002<-HG002_rep %>% read.table(header=FALSE) %>% as_tibble() %>% 
  select(V1,V4,V8,V9) %>% rename(Chr=V1, score=V4, Repeat=V8, PrintName=V9) %>%
  group_by(Repeat) %>% mutate(count=n(), Group="Full repeat", Type="HG002")
rep_tib_chm13<-chm13_rep %>% read.table(header=FALSE) %>% as_tibble() %>% 
  select(V1,V4,V8,V9) %>% rename(Chr=V1, score=V4, Repeat=V8, PrintName=V9) %>% 
  group_by(Repeat) %>% mutate(count=n(), Group="Full repeat", Type="CHM13")
# Tibble with scores - only G4s
g4_tib_HG002<-HG002_g4 %>% read.table(header=FALSE) %>% as_tibble() %>% 
  select(V1,V4,V9,V10) %>% rename(Chr=V1, score=V4, Repeat=V9,PrintName=V10)  %>%
  group_by(Repeat) %>% mutate(count=n(), Group="G4 in repeat", Type="HG002")
g4_tib_chm13<-chm13_g4 %>% read.table(header=FALSE) %>% as_tibble() %>% 
  select(V1,V4,V9,V10) %>% rename(Chr=V1, score=V4, Repeat=V9,PrintName=V10)  %>%
  group_by(Repeat) %>% mutate(count=n(), Group="G4 in repeat", Type="CHM13")
# Combine
METHDATA<- rep_tib_HG002 %>% full_join(g4_tib_HG002) %>% full_join(rep_tib_chm13) %>% 
  full_join(g4_tib_chm13)  %>% filter(!grepl("COMP",Repeat)) %>%
  mutate(PrintName = sub("SAT", "Sat", PrintName))
METHDATA$PrintName<-factor(METHDATA$PrintName, levels=c("SVA","rRNA","ACRO1","GSat","GSatII","GSatX","HSat5","LSAU","MSR1","Sat-VAR","SST1","TAR1"))
METHDATA$Group<-factor(METHDATA$Group, levels=c("Full repeat", "G4 in repeat"))

# REPEAT GROUP DATA 
GROUP <- groupRep_file %>% read.table(header=TRUE) %>% as_tibble() %>% 
  pivot_longer(-c(Repeat,Species), names_to="Motif") %>% 
  mutate(Motif = str_replace_all(Motif, c('GQ' = 'G4'))) %>% 
  mutate(Name=case_when(Species=="bonobo" ~ "Bonobo",
                        Species=="chimp" ~ "Chimpanzee",
                        Species=="human" ~ "Human",
                        Species=="gorilla" ~ "Gorilla",
                        Species=="borang" ~ "B. orangutan",
                        Species=="sorang" ~ "S. orangutan",
                        Species=="siamang" ~ "Siamang")) %>% 
  mutate(logval=log2(value)) 
  

GROUP$Motif <- factor(GROUP$Motif, levels=c("APR", "DR", "STR", "IR", "MR", "TRI", "G4", "Z"))
GROUP$Repeat <- factor(GROUP$Repeat, levels=c("TEs", "Satellites", "RNA", "Other"))
GROUP$Name <- factor(GROUP$Name, levels=c("Bonobo", "Chimpanzee", "Human", "Gorilla", "B. orangutan", "S. orangutan", "Siamang"))

################################################################################
# PLOT PART1: GROUP REPEAT ENRICHMENT

scalestart=min(GROUP %>% filter(value>0) %>%select(logval))
scaleend=max(GROUP$logval)
scalebreak=-1*scalestart/(scaleend-scalestart)

p1<-ggplot(GROUP, aes(Name, fct_rev(Motif))) +
  geom_tile(aes(fill=logval), color = "white",lwd = 0.3,linetype = 1) +
  facet_grid(~Repeat, space="free", scales="free_x") +
  scale_fill_gradientn(name="log2(Enrichment)",
                       colors=c("#075AFF", "white","#FF0000"),
                       values=c(0, scalebreak, 1), na.value="#075AFF")+
  labs(x="", y="") +
  theme(axis.text.x = element_text(angle = 55, vjust = 1, hjust=1, size=8),
        plot.margin = margin(t=0,r=0,b=0,l=0, unit="in"),
        legend.position="right",
        legend.text=element_text(size=8),
        legend.key.width  = unit(0.7, "lines"),
        legend.key.height = unit(1.5, "lines"),
         legend.title=element_text(size=10, vjust=1, angle=90),
        legend.title.position="left",
        panel.spacing = unit(2, "lines"),
        legend.margin = margin(t=0,r=0,b=0,l=0, unit="pt"),
        plot.title = element_text(size=18),
        strip.text = element_text(colour = 'black', size=11, face="bold"),
        panel.background = element_rect(fill = 'white', colour="white"))+
  scale_y_discrete(position="left")+
  labs(title = bold('A') ~ '')
p1

################################################################################
# PLOT PART2: REPEAT ENRICHMENT HUMAN

p2<-ggplot(REPDATA, aes(PrintName, forcats::fct_rev(Motif))) +
  geom_tile(aes(fill=logval), color = "white",lwd = 0.3,linetype = 1) +
  facet_grid(~Class, space="free", scales="free_x") +
  geom_point(data=REPDATA %>% filter(Mark=="Yes"), aes(x=PrintName, y=9), color="black", fill="black", inherit.aes = FALSE, pch=25) +
  scale_fill_gradientn(name="log2(Enrichment)",
                       colors=c("#075AFF", "white","#FF0000"),
                       values=c(0, 10/16, 1), na.value="#075AFF")+
  labs(x="", y="") +
  theme(axis.text.x = element_text(angle = 55, vjust = 1, hjust=1, size=8),
        plot.margin = margin(t=0,r=0,b=0,l=0, unit="pt"),
        legend.position="right",
        legend.text=element_text(size=8),
        legend.title=element_text(size=10, vjust=1, angle=90),
        legend.title.position="left",
        legend.key.width  = unit(0.7, "lines"),
        legend.key.height = unit(1.5, "lines"),
        legend.margin = margin(t=0,r=0,b=0,l=0, unit="pt"),
        plot.title = element_text(size=18),
        strip.text = element_text(colour = 'black', size=11, face="bold"),
        panel.background = element_rect(fill = 'white', colour="white"))+
  scale_y_discrete(position="left")+
  labs(title = bold('B') ~ '')
p2


################################################################################
# PLOT PART3: METHYLATION IN SATELLITES 

p3<-ggplot(METHDATA, aes(x=score, fill=Group, color=Group)) + 
  geom_density(alpha=0.7) +
  scale_fill_manual(values=c("darkgrey", "#DB5829"))+
  scale_color_manual(values=c("darkgrey", "#DB5829"))+
  ggh4x::facet_grid2(Type ~ PrintName, scales = "free_y", independent = "y") +
  theme(panel.background = element_rect(fill = 'white', colour="white"),
        panel.grid.major = element_line(colour = 'beige'),
          legend.position="bottom",
      legend.margin = margin(t=0,r=0,b=0,l=0, unit="pt"),
        legend.title=element_blank(),
          axis.ticks.x=element_blank(), 
         axis.ticks.y=element_blank(),
        strip.background = element_rect(fill='lightgray'),
        plot.title = element_text(size=18),
        axis.title=element_text(size=12),
        strip.text = element_text(colour = 'black', size=10, face="bold"))+
  scale_x_continuous("Methylation score", breaks=c(0,0.5,1), labels=c(0,0.5,1), expand = c(0,0))+
  scale_y_continuous("Density", expand = c(0,0))+
  labs(title = bold('C') ~ '')
p3


################################################################################
# COMBINE REPEAT ENRICHMENT PLOT WITH METHYATION PLOT

# ABC together 
pcomb <- p1 + p2 + p3 + plot_layout(heights=c(1,1,2))
pcomb
outfile=paste(plotdir,"Fig6ABC_Repeat_Methylation.png", sep="")
ggsave(outfile,plot = pcomb,scale = 1,dpi = 600,limitsize = TRUE,width=11,height=9)
outfile=paste(plotdir,"Fig6ABC_Repeat_Methylation.pdf", sep="")
ggsave(outfile,plot = pcomb,scale = 1,dpi = 600,limitsize = TRUE,width=12,height=9)
